File: go/pkg/crypto/aes-gcm.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"io"
)

func EncryptAESGCM(message string, key []byte) (string, error) {
	if len(key) != 16 && len(key) != 24 && len(key) != 32 {
		return "", fmt.Errorf("invalid key size: %d", len(key))
	}

	byteMsg := []byte(message)
	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("could not create new cipher: %v", err)
	}

	// Generate a random nonce
	nonce := make([]byte, 12)
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return "", fmt.Errorf("could not generate nonce: %v", err)
	}

	// Create AEAD cipher
	aead, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("could not create GCM: %v", err)
	}

	// Encrypt and authenticate
	ciphertext := aead.Seal(nonce, nonce, byteMsg, nil)
	return base64.StdEncoding.EncodeToString(ciphertext), nil
}

func DecryptAESGCM(message string, key []byte) (string, error) {
	ciphertext, err := base64.StdEncoding.DecodeString(message)
	if err != nil {
		return "", fmt.Errorf("could not base64 decode: %v", err)
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("could not create new cipher: %v", err)
	}

	aead, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("could not create GCM: %v", err)
	}

	if len(ciphertext) < 12 {
		return "", fmt.Errorf("ciphertext too short")
	}

	nonce := ciphertext[:12]
	ciphertext = ciphertext[12:]

	// Decrypt and verify
	plaintext, err := aead.Open(nil, nonce, ciphertext, nil)
	if err != nil {
		return "", fmt.Errorf("could not decrypt: %v", err)
	}

	return string(plaintext), nil
}

File: ts/src/crypto/aes-gcm.ts
const crypto = globalThis.crypto;

export async function encryptAESGCM(
  message: string,
  key: Uint8Array,
  nonce?: Uint8Array
): Promise<Uint8Array> {
  // Validate key size
  if (![16, 24, 32].includes(key.length)) {
    throw new Error(`Invalid key size: ${key.length}`);
  }

  // Validate nonce if provided
  if (nonce && nonce.length !== 12) {
    throw new Error("Nonce must be 12 bytes");
  }

  // Convert message to bytes
  const msgBuffer = new TextEncoder().encode(message);

  // Generate random nonce if not provided
  if (!nonce) {
    nonce = crypto.getRandomValues(new Uint8Array(12));
  }

  // Import key
  const cryptoKey = await crypto.subtle.importKey(
    "raw",
    key,
    { name: "AES-GCM" },
    false,
    ["encrypt"]
  );

  // Encrypt
  const ciphertext = await crypto.subtle.encrypt(
    {
      name: "AES-GCM",
      iv: nonce,
      tagLength: 128,
    },
    cryptoKey,
    msgBuffer
  );

  // Combine nonce and ciphertext
  const result = new Uint8Array(nonce.length + ciphertext.byteLength);
  result.set(nonce);
  result.set(new Uint8Array(ciphertext), nonce.length);

  return result;
}

export async function decryptAESGCM(
  encrypted: Uint8Array,
  key: Uint8Array
): Promise<Uint8Array> {
  try {
    if (encrypted.length < 28) {
      // 12 (nonce) + 16 (minimum tag size)
      throw new Error("Ciphertext too short");
    }

    // Extract nonce and ciphertext
    const nonce = encrypted.subarray(0, 12);
    const ciphertext = encrypted.subarray(12);

    // Import key
    const cryptoKey = await crypto.subtle.importKey(
      "raw",
      key,
      { name: "AES-GCM" },
      false,
      ["decrypt"]
    );

    // Decrypt
    const plainBuffer = await crypto.subtle.decrypt(
      {
        name: "AES-GCM",
        iv: nonce,
        tagLength: 128,
      },
      cryptoKey,
      ciphertext
    );

    return new Uint8Array(plainBuffer);
  } catch (err: any) {
    throw new Error(`Decryption failed: ${err.message}`);
  }
}
